src/dbus_api/types.rs

```diff
@@ -0,0 +1,121 @@
+        DEV_NOTFOUND,
+        FILESYSTEM_NOTFOUND,
+        IO_ERROR,
+        LIST_FAILURE,
+        INTERNAL_ERROR,
+        NIX_ERROR,
+        NO_POOLS,
+        NOTFOUND,
+        NULL,
+        NULL_NAME,
+        POOL_NOTFOUND,
+    }
+}
+
+/// Get the u16 value of this ErrorEnum constructor.
+impl From<ErrorEnum> for u16 {
+    fn from(e: ErrorEnum) -> u16 {
+        e as u16
+    }
+}
+
+impl ErrorEnum {
+    pub fn get_error_string(&self) -> &str {
+        match *self {
+            ErrorEnum::OK => "Ok",
+            ErrorEnum::ERROR => "A general error happened",
+            ErrorEnum::NULL => "Null parameter was supplied",
+            ErrorEnum::NOTFOUND => "Not found",
+            ErrorEnum::POOL_NOTFOUND => "Pool not found",
+            ErrorEnum::FILESYSTEM_NOTFOUND => "Filesystem not found",
+            ErrorEnum::CACHE_NOTFOUND => "Cache not found",
+            ErrorEnum::BAD_PARAM => "Bad parameter",
+            ErrorEnum::DEV_NOTFOUND => "Dev not found",
+            ErrorEnum::ALREADY_EXISTS => "Already exists",
+            ErrorEnum::NULL_NAME => "Null name supplied",
+            ErrorEnum::NO_POOLS => "No pools",
+            ErrorEnum::LIST_FAILURE => "List operation failure.",
+            ErrorEnum::INTERNAL_ERROR => "Internal error",
+            ErrorEnum::IO_ERROR => "IO error during operation.",
+            ErrorEnum::NIX_ERROR => "System error during operation.",
+            ErrorEnum::BUSY => "Operation can not be performed at this time",
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum DeferredAction {
+    Add(ObjectPath<MTFn<TData>, TData>),
+    Remove(String),
+}
+
+#[derive(Debug, Clone)]
+pub struct DbusContext {
+    pub next_index: Rc<Cell<u64>>,
+    pub pools: Rc<RefCell<BidirMap<String, String>>>,
+    pub engine: Rc<RefCell<Box<Engine>>>,
+    pub action_list: Rc<RefCell<Vec<DeferredAction>>>,
+    pub filesystems: Rc<RefCell<BidirMap<String, (String, String)>>>,
+}
+
+impl DbusContext {
+    pub fn new(engine: Rc<RefCell<Box<Engine>>>) -> DbusContext {
+        DbusContext {
+            action_list: Rc::new(RefCell::new(Vec::new())),
+            engine: engine.clone(),
+            filesystems: Rc::new(RefCell::new(BidirMap::new())),
+            next_index: Rc::new(Cell::new(0)),
+            pools: Rc::new(RefCell::new(BidirMap::new())),
+        }
+    }
+
+    /// Generates a new id for object paths.
+    /// It is assumed that, while Stratisd is running, it will never generate
+    /// more than 2^64 object paths. If it turns out that this is a bad
+    /// assumption, the solution is to use unbounded integers.
+    pub fn get_next_id(&mut self) -> u64 {
+        self.next_index.set(self.next_index.get() + 1);
+        self.next_index.get()
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct TData;
+impl DataType for TData {
+    type ObjectPath = DbusContext;
+    type Property = ();
+    type Interface = ();
+    type Method = ();
+    type Signal = ();
+}
```

Comment:
I promise to add more stuff to it...in fact PR #193 already does, that stuff just isn't ready yet.